File: TypeScript/BestTimeToBuyAndSellStockIII.ts
function maxProfit(prices: number[]): number {
    let oneBuyOneSell: number = 0;
    let twoBuyTwoSell: number = 0;
    let oneBuy: number = Number.POSITIVE_INFINITY;
    let twoBuy: number = Number.POSITIVE_INFINITY;
    for(let i = 0; i < prices.length; i++) {
        const price: number = prices[i];
        oneBuy = Math.min(oneBuy, price);
        oneBuyOneSell = Math.max(oneBuyOneSell, price - oneBuy);
        twoBuy = Math.min(twoBuy, price - oneBuyOneSell);
        twoBuyTwoSell = Math.max(twoBuyTwoSell, price - twoBuy);
    }
    return twoBuyTwoSell;
};

File: Java/BestTimeToBuyAndSellStockIII.java
class Solution {
    public int maxProfit(int[] prices) {
        int oneBuyOneSell = 0;
        int twoBuyTwoSell = 0;
        int oneBuy = Integer.MAX_VALUE;
        int twoBuy = Integer.MAX_VALUE;
        int price = 0; 
        for(int i = 0; i < prices.length; i++) {
            price = prices[i];
            oneBuy = Math.min(oneBuy, price);
            oneBuyOneSell = Math.max(oneBuyOneSell, price - oneBuy);
            twoBuy = Math.min(twoBuy, price - oneBuyOneSell);
            twoBuyTwoSell = Math.max(twoBuyTwoSell, price - twoBuy);
        }
        return twoBuyTwoSell;
    }
}